### PR TITLE
[FIX] mass_mailing: More accurate next_departure taking into account cron.triggers

### PR DESCRIPTION
**Pull request by jorv-odoo**
When the user schedules a datetime for the emailing campaign, it creates
a ir.cron.trigger entry that should run the Mail Marketing: Process queue.
This fix tries to take this into account to calculate a next_departure
datetime closer to the true moment job will be run.

When the schedule datetime is between now() and the next call datetime
for the cron job (nextcall field), current logic will always chose
nextcall given the use of max(). This can be confusing for the end-user
as the information banner in the form view will indicate the nextcall date,
which in some cases can be way farther in the futur.

While a ir.cron.trigger is not a 100% guarantee that the cron job will
run at that specific time, it will give better feedback to the user for
when he should expected his mailing campaign to be send out.

Main changes:

- simplified logic in `_compute_next_departure` to take into account
the implicit creation of cron.triggers (methods `action_launch` and
`action_schedule`)

- added test_mailing_next_departure to simulate "Send", "Cancel" and
"Schedule Date" buttons on mass_mailing form view. Check ir.cron.trigger
model for presence of triggers for cron job "Mail Marketing: Process queue"
and assert if cron.trigger was created as expected

opw-3137819
original commit https://github.com/odoo/odoo/commit/b98494dca7ea5499d29139297308071f6f15bedc
some trailing whitespace got linted from commit [4cebf5](https://github.com/odoo/odoo/commit/4cebf53aa89c21ef255f5d5da450fdf98c4c9ac1)

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/mass_mailing/models/mailing.py [hunk 1]
```diff
@@ -266,7 +266,10 @@ def _compute_next_departure(self):
         for mass_mailing in self:
             if mass_mailing.schedule_date:
                 schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                if fields.Datetime.now() < schedule_date < cron_time:
+                    mass_mailing.next_departure = schedule_date
+                else:
+                    mass_mailing.next_departure = max(schedule_date, cron_time)
```
Comment (thiryr) [inline addons/mass_mailing/models/mailing.py:272]: Neat, didn't know about comparison chaining.

Not saying we can't, but isn't this just lying to users to make the interface feedback less confusing?

From the PR message I get the impression this is meant to be more accurate, but if the cron runs after the scheduled date, then that's when the mailing would be sent. So it's less accurate, right?
Unless I'm missing something

**Review by thiryr — COMMENTED** [review]
Howdy :wave: 

Small comment because I'm a bit confused about the purpose :)

File: addons/mass_mailing/models/mailing.py (hunk 1 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:272]: Hey thanks for taking a look!

Since Odoo 15 ( might even been added in 14.2) we have the cron triggers which allow for a cron job to be triggered before its saved nextcall value.

From my naive understanding, when the shared cron workers on a SAAS server show up to that specific DB, they check if there is a trigger with a datetime before the `nextcall`, and runs the job early. There seems to be a time buffer where the job might start slightly later (the workers might be busy handling other jobs), but it should be close to the datetime of the trigger.

The default value for the “Mail Marketing: Process queue” is once every 1 days. so lets say:
- User clicks send now or schedules the mailing list for FEB10 9H10
- Trigger is created for that datetime
- Nextcall for “Mail Marketing: Process queue” happens to be FEB11 10H00

The user will see a banner with "This mailing is planned for FEB11 10H00". This might be confusing and lead the user to think that there was a mistake. Worse in my support ticket case he taught it was a good idea to go into the technical menu and run  “Mail Marketing: Process queue” manually, which lead to a bug where the mailing campaign was sending mails twice to the same recipients (out of scope of this precise PR).

So while the cron trigger does not guarantee that the job will run at that precise moment, is is closer to the truth than informing the client that the mails will go out 1 day later vs an hour later (based on the trigger).

Comment (thiryr) [reply addons/mass_mailing/models/mailing.py:272]: Oh wow, I see. Thanks for the explanation, I naively assumed nextCall accounted for these.
Indeed this looks like it was not updated when the time-scheduled cron job was added to mailings.

In that case why not do something like max(fields.Datetime.now(), schedule_date)? If I schedule in the past, it will be processed 'instantly' (on the next pass of a worker on this DB), and otherwise it will always be triggered around the scheduled date with the new cron trigger, right?

Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:272]: When you breakpoint the `def _compute_next_departure(self):` and then use the front-end to schedule a "draft" mailing list, it runs a few times over that method.
- First run, since `mass_mailing.schedule_date` is not set (FALSE), it defaults to `mass_mailing.next_departure = cron_time`
- Further down stream, `action_schedule_date` takes the user input for a date, sets `mass_mailing.schedule_date` and creates the cron.trigger at the same date.time (using `self.mass_mailing_id.action_put_in_queue` in [mailing_mailing_schedule_date.py](https://github.com/odoo/odoo/blob/16.0/addons/mass_mailing/wizard/mailing_mailing_schedule_date.py))
- It comes back to `_compute_next_departure` and triggers the `if mass_mailing.schedule_date` condition. Only here my changes apply for the described edge case.

So your simplification would be to assume that the cron trigger (implicitly created) time is more relevant than cron time and just replace `max(schedule_date, cron_time)` with `max(fields.Datetime.now(), schedule_date)` ?

I can get behind that, since the behavior since Odoo 15 should be the one described above. I just hope that I am not missing some unforeseen consequences of that change.

Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py]: Tested `max(fields.Datetime.now(), schedule_date) ` : behaves as expected in my local test and seems a clean solution.
Pushed a new commit. Do you have any other input or ideas?

File: addons/mass_mailing/models/mailing.py [hunk 2]
```diff
@@ -260,15 +260,30 @@ def _compute_statistics(self):
...
+        cron_job = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo()
         str2dt = fields.Datetime.from_string
-        cron_time = str2dt(cron_next_call)
+        cron_time = str2dt(cron_job.nextcall)
+        cron_id = cron_job.id
+
+        # The cron worker usually clean up old cron.triggers after finishing their
+        # jobs, i.e older than now() thus fetched recordset should not be to large
+        now_datetime_buffer = fields.Datetime.now() + relativedelta(minutes=1)
+        now_datetime_str = fields.Datetime.to_string(now_datetime_buffer)
```
Comment (thiryr) [inline addons/mass_mailing/models/mailing.py:271]: I believe the ORM will do this for you?

File: addons/mass_mailing/models/mailing.py [hunk 3]
```diff
@@ -260,15 +260,30 @@ def _compute_statistics(self):
...
+        cron_triggers = self.env['ir.cron.trigger'].sudo().search(['&', ('cron_id', '=', cron_id), ('call_at', '<', now_datetime_str)])
+        most_recent_trigger_datetime = max([str2dt(tr.call_at) for tr in cron_triggers]) if cron_triggers else False
```
Comment (thiryr) [inline addons/mass_mailing/models/mailing.py:273]: you can use `order='call_at desc'` and `limit=1` as parameters for `search` so that all that is done by Postgres

File: addons/mass_mailing/models/mailing.py [hunk 4]
```diff
@@ -260,15 +260,30 @@ def _compute_statistics(self):
...
         for mass_mailing in self:
             if mass_mailing.schedule_date:
                 schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                mass_mailing.next_departure = max(schedule_date, fields.datetime.now())
```
Comment (thiryr) [inline addons/mass_mailing/models/mailing.py:277]: Don't we also want to use the next trigger now that we know when it will happen?

File: addons/mass_mailing/models/mailing.py (hunk 3 above)
Comment (thiryr) [reply addons/mass_mailing/models/mailing.py:273]: Also I think you meant this, right?
```suggestion
        cron_triggers = self.env['ir.cron.trigger'].sudo().search(['&', ('cron_id', '=', cron_id), ('call_at', '>', now_datetime_str)])
        most_recent_trigger_datetime = max([str2dt(tr.call_at) for tr in cron_triggers]) if cron_triggers else False
```

File: addons/mass_mailing/models/mailing.py [hunk 5]
```diff
@@ -260,15 +260,30 @@ def _compute_statistics(self):
...
             else:
-                mass_mailing.next_departure = cron_time
+                # We checked for cron.trigger relating to the cron_id
+                # If at least one more recent than "now" exists we can reasonably
+                # expect that the cron worker will execute this mass_mailing job
+                # the next time it wakes up
+                if most_recent_trigger_datetime:
+                    mass_mailing.next_departure = max(most_recent_trigger_datetime, fields.datetime.now())
```
Comment (thiryr) [inline addons/mass_mailing/models/mailing.py:284]: I believe the search domain already guarantees that one will be greated than the other, right?

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 6]
```diff
@@ -31,3 +31,77 @@ def test_mailing_schedule_date(self):
...
+
+        # set a schedule date after cron_job_nextcall
+        schedule_date_user = cron_job_nextcall + relativedelta(days=50)
+        wizard_form.schedule_date = schedule_date_user
+        wizard = wizard_form.save()
+        wizard.action_schedule_date()
+
+        # assert if schedule_date/departure is after cron_job_nextcall
+        test_after_cron_job_time = mailing.next_departure > cron_job_nextcall
+        self.assertEqual(test_after_cron_job_time, True)
```
Comment (thiryr) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:103]: `assertTrue` exists ^^
(same above)

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 7]
```diff
@@ -31,3 +31,77 @@ def test_mailing_schedule_date(self):
         self.assertEqual(mailing.schedule_date, datetime(2021, 4, 30, 9, 0))
         self.assertEqual(mailing.schedule_type, 'scheduled')
         self.assertEqual(mailing.state, 'in_queue')
+
+    def test_mailing_next_departure(self):
+        # test if mailing.mailing.next_departure is correctly set taking into account
+        # presence of implicitly created cron triggers (since odoo v15). These should
+        # launch cron job before its schedule nextcall datetime (if scheduled_date < nextcall)
+
+        # case were user click on "Send" button (action_launch)
```
Comment (thiryr) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:40]:
```suggestion
        # case where user click on "Send" button (action_launch)
```

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 8]
```diff
@@ -31,3 +31,77 @@ def test_mailing_schedule_date(self):
...
+        # case where client uses schedule wizard to chose a date between now and cron.job nextcall
+        # create a schedule date wizard
+        wizard_form = Form(
+            self.env['mailing.mailing.schedule.date'].with_context(default_mass_mailing_id=mailing.id))
+
+        # set a schedule date in near future
+        schedule_date_user = datetime.now() + relativedelta(minutes=50)
+        wizard_form.schedule_date = schedule_date_user
+        wizard = wizard_form.save()
+        wizard.action_schedule_date()
```
Comment (thiryr) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:80]: I would just create a new mailing with the proper state. This code seems irrelevant to what is being tested imo.

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 9]
```diff
@@ -31,3 +31,77 @@ def test_mailing_schedule_date(self):
...
+        mailing.action_cancel()
+        # # create a schedule date wizard
+        wizard_form = Form(
+            self.env['mailing.mailing.schedule.date'].with_context(default_mass_mailing_id=mailing.id))
+
+        # set a schedule date after cron_job_nextcall
+        schedule_date_user = cron_job_nextcall + relativedelta(days=50)
+        wizard_form.schedule_date = schedule_date_user
+        wizard = wizard_form.save()
+        wizard.action_schedule_date()
```
Comment (thiryr) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:99]: Same as above. Unless relevant I would just create a new mailing with the proper state.

(Also at that point it could just be 3 tests, but I guess that's more a matter of taste)

**Review by thiryr — COMMENTED** [review]
Hey there, I have some relatively minor comments

Overall pretty nice

File: addons/mass_mailing/models/mailing.py (hunk 2 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:271]: (Thanks for the overall input, will comment as I go along)

That was my understanding, but since the original method definition used the `to_string` approach, I didn't know if I should keep the same approach or try to let the ORM do the work. Will try to clean it up a bit.

File: addons/mass_mailing/models/mailing.py (hunk 3 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:273]: From my debugging session, the compute gets called before the trigger is created, then after it's created. So the trigger call_at date should be slightly younger than now (added a buffer of 1 minute just to be sure). So the less than logic holds up, since what we want is to be sure that there is at least one trigger younger that now for mailing object where `schedule_date = False`.

File: addons/mass_mailing/models/mailing.py (hunk 4 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:277]: As described before, there might be a slight timeshift (milliseconds at runtime?) between when the trigger is set and `next_departure` is set. I guess the time difference is negligible, especially at the front-end facing time precision. We could just set it to `schedule_date`, since the trigger is created based on `schedule_date` in that scenario.

File: addons/mass_mailing/models/mailing.py (hunk 5 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:284]: Idem as before, we could remove the max call and just set it to most_recent_trigger_datetime as the time difference with now should be negligible.

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py (hunk 6 above)
Comment (jorv-odoo) [reply addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:103]: Tests are new for me, thanks for teaching me new stuff ;)

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py (hunk 9 above)
Comment (jorv-odoo) [reply addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:99]: I indeed hesitate to split it up into three.
Arguments for one test:
- We test if the user behavior of clicking "Send" > "Cancel" > "Schedule date" correctly resets the states/fields and correctly schedules the mailing list for two scenarios (before nextcall cron job and case of after cron job nextcall)

- For the `Form` class call, I followed the logic of `test_mailing_schedule_date` (same file with a test committed 23 months ago). Since schedule_date usually calls the calendar wizard, I guess it's closer to "real" user behavior.

Let me know if you prefer splitting it up or cleaning it up based on the above observations!

File: addons/mass_mailing/models/mailing.py (hunk 3 above)
Comment (thiryr) [reply addons/mass_mailing/models/mailing.py:273]: I'm not sure I get the point, here's my understanding:

- I give a schedule_date to a mailing
- Whenever the cron is triggered, if the schedule_date is past, the mailing is sent
- If my schedule date is before the next trigger, then what is relevant is actually the trigger.
- So for any mailing the next departure should be `max(min(next_trigger, next_cron), schedule_date)`
 with `next_trigger` being the next trigger after `max(now, schedule_date)`
 -> note that this is idealized as it could get a bit expensive to check the next trigger after each schedule_date if we need to look at all the triggers, and it probably won't bring much value.

That specific compute will run every time we fetch this record so it shouldn't be a big deal. But if there really is a microsecond discrepency issue, we can just add some negative padding to 'now' to be sure right? Though I'm not a fan.

Let me know if I got something wrong or if it's not clear :)

**Review by thiryr — COMMENTED** [review]
Response

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 10]
```diff
@@ -31,3 +31,77 @@ def test_mailing_schedule_date(self):
         self.assertEqual(mailing.schedule_date, datetime(2021, 4, 30, 9, 0))
         self.assertEqual(mailing.schedule_type, 'scheduled')
         self.assertEqual(mailing.state, 'in_queue')
+
+    def test_mailing_next_departure(self):
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:35]: Note that you may use freezetime to use fake datetimes, and notably force "now" to be whatever you want. It allows notably to run some schedulers while faking the current datetime and check that conditions are well respected. This generally leads to more readable code :)

File: addons/mass_mailing/models/mailing.py (hunk 3 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:273]: This is indeed a way we could simplify it.
If we fetch the most recent trigger (younger than `now`), to treat all possible cases:
- user schedules a date in the past `we are 10 FEB, user puts 01FEB` (should this even be possible? because right now it is) : trigger is created for 01FEB. Since `schedule_date` gets set, we go to `max(mass_mailing.schedule_date, fields.datetime.now())`, as the trigger is way in the past, it should wake the the worker execute it now [ OUTPUT : fields.datetime.now()]
- user presses "send":  `schedule_date` stays `False` and trigger for `datetime.now()` is created. Here we could just go for `mass_mailing.next_departure = most_recent_trigger_datetime` [ OUTPUT : most_recent_trigger_datetime]
- user schedules a date in the future: `schedule_date`  is set and trigger gets created in the future (after `now`). Here `mass_mailing.next_departure = max(mass_mailing.schedule_date, fields.datetime.now())` kicks in again and should give the close enough next_departure  [ OUTPUT : mass_mailing.schedule_date ]

(Not disagreeing with you but just giving my summary of the use cases. If I missed something point it out :)

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py (hunk 10 above)
Comment (jorv-odoo) [reply addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:35]: Thanks for the info. Reworking the test and using your input!

File: addons/mass_mailing/models/mailing.py [hunk 11]
```diff
@@ -260,15 +260,34 @@ def _compute_statistics(self):
...
-                mass_mailing.next_departure = cron_time
+                # Schedule_date should is only False if schedule_type = "now" or
+                # mass_mailing is canceled.
+                # We checked for cron.triggers relating to the cron_id
+                # so we can reasonably expect that the cron worker will
+                # execute this mass_mailing job the next time it wakes up
+                if most_recent_trigger_datetime:
+                    mass_mailing.next_departure = most_recent_trigger_datetime
+                else:
+                    mass_mailing.next_departure = cron_time
```
Comment (jorv-odoo) [inline addons/mass_mailing/models/mailing.py:290]: Simpliefied code based on feedback.

Concerning `cron_time`: as we now rely on cron trigger times, I guess we could get rid of that "fail safe"? Normally based on current code behavior, we should always have at least one trigger created thus outputting either the schedule_date/trigger_datetime or `now`, whichever is more relevant. 

Let me know if we should remove it, as I mostly left it based on previous authors' commit and in case I was missing some unforeseen edge case.

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py (hunk 10 above)
Comment (jorv-odoo) [reply addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:36]: Added freeze_time approach on most recent commit!

File: addons/mass_mailing/models/mailing.py [hunk 12]
```diff
@@ -260,15 +260,34 @@ def _compute_statistics(self):
...
+            order='call_at desc', limit=1)
+        most_recent_trigger_datetime = cron_triggers.call_at
         for mass_mailing in self:
             if mass_mailing.schedule_date:
-                schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                mass_mailing.next_departure = max(mass_mailing.schedule_date, fields.datetime.now())
             else:
-                mass_mailing.next_departure = cron_time
+                # Schedule_date should is only False if schedule_type = "now" or
```
Comment (thiryr) [inline addons/mass_mailing/models/mailing.py:282]:
```suggestion
                # Schedule_date should only be False if schedule_type = "now" or
```

**Review by thiryr — COMMENTED** [review]
Ok for me, no more comments
(not approving as TDE should probably have a last look on this)

Thanks for being on top of this :)

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 13]
```diff
@@ -31,3 +32,65 @@ def test_mailing_schedule_date(self):
...
+        # test if mailing.mailing.next_departure is correctly set taking into account
+        # presence of implicitly created cron triggers (since odoo v15). These should
+        # launch cron job before its schedule nextcall datetime (if scheduled_date < nextcall)
+
+        cron_job = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue').sudo()
+        cron_job.write({'nextcall' : datetime(2023, 2, 18, 9, 0)})
+
+        # case were user click on "Send" button (action_launch)
```
Comment (jorv-odoo) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:44]:
```suggestion
        # case where user click on "Send" button (action_launch)
```

File: addons/mass_mailing/models/mailing.py (hunk 11 above)
Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing.py:290]: I think having a fail safe is always a good idea :) we never know what people did with their data ... does not cost much.

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 14]
```diff
@@ -31,3 +32,65 @@ def test_mailing_schedule_date(self):
         self.assertEqual(mailing.schedule_date, datetime(2021, 4, 30, 9, 0))
         self.assertEqual(mailing.schedule_type, 'scheduled')
         self.assertEqual(mailing.state, 'in_queue')
+
+    def test_mailing_next_departure(self):
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:36]: Linting: could you order test methods alphabetically (aka defining it before test_mailing_schedule_date) ? Reason is that as they are executed alphabetically in a given class, better follow the same order in definition so that execution matches file order.

File: addons/mass_mailing/models/mailing.py [hunk 15]
```diff
@@ -260,15 +260,34 @@ def _compute_statistics(self):
...
+            '&',
+            ('cron_id', '=', cron_id),
+            ('call_at', '<', now_datetime_buffer)],
+            order='call_at desc', limit=1)
+        most_recent_trigger_datetime = cron_triggers.call_at
         for mass_mailing in self:
             if mass_mailing.schedule_date:
-                schedule_date = str2dt(mass_mailing.schedule_date)
-                mass_mailing.next_departure = max(schedule_date, cron_time)
+                mass_mailing.next_departure = max(mass_mailing.schedule_date, fields.datetime.now())
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:280]: Does not seem this case is tested ?

File: addons/mass_mailing/models/mailing.py [hunk 16]
```diff
@@ -260,15 +260,34 @@ def _compute_statistics(self):
...
             else:
-                mass_mailing.next_departure = cron_time
+                # Schedule_date should only be False if schedule_type = "now" or
+                # mass_mailing is canceled.
+                # We checked for cron.triggers relating to the cron_id
+                # so we can reasonably expect that the cron worker will
+                # execute this mass_mailing job the next time it wakes up
+                if most_recent_trigger_datetime:
+                    mass_mailing.next_departure = most_recent_trigger_datetime
```
Comment (tde-banana-odoo) [inline addons/mass_mailing/models/mailing.py:288]: Looking at tests, unsure this is effectively tested ?

**Review by tde-banana-odoo — COMMENTED** [review]
Fix LGTM, thanks for the PR :) just some comments, and it seems some use cases are not completely covered by unit tests (but still did not have my :coffee: so maybe I am wrong).

File: addons/mass_mailing/models/mailing.py (hunk 15 above)
Comment (awa-odoo) [inline addons/mass_mailing/models/mailing.py:280]: Joining in on this, for this use case we just assume that there will be a cron trigger created at schedule time and we do not check its existence?

Seems a bit simple, but I'm guessing all the flows cover that? (Putting in queue / editing scheduled date / ...) ?
Makes me wonder why we are checking that it exists when sending immediately though.

Trying to fetch all cron triggers would be more complex, as you could have "self" containing multiple mailings with different schedule date.

Maybe I missed something.

Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:280]: This case kicks in if the client uses the schedule wizard to point to a date in the past (before `now()`).
As of now there is nothing preventing them from doing this, so I don't know if this is a bug or a "feature".
SO if we accept that the user can schedule a date in the past, I probably should add a test for that case!

Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:280]: Thx for the input.

As you guessed I am working with the implicit nature for the flow here
```   
    def action_launch(self):
        self.write({'schedule_type': 'now'})
        return self.action_put_in_queue()

    def action_schedule(self):
        self.ensure_one()
        if self.schedule_date:
            return self.action_put_in_queue()
        else:
            action = self.env["ir.actions.actions"]._for_xml_id("mass_mailing.mailing_mailing_schedule_date_action")
            action['context'] = dict(self.env.context, default_mass_mailing_id=self.id)
            return action

    def action_put_in_queue(self):
        self.write({'state': 'in_queue'})
        cron = self.env.ref('mass_mailing.ir_cron_mass_mailing_queue')
        cron._trigger(
            schedule_date or fields.Datetime.now()
            for schedule_date in self.mapped('schedule_date')
        )
```

The thing is that `mass_mailing.schedule_date` only gets set when using the schedule wizard, i.e. stays False when using "Send" button (which goes directly to `.action_launch`) .

Now for this code line in question, we could fetch the trigger, but the issue is that if the user schedules a date way in the future ( after now or nextcall), we would need to add an extra `search()` matching the datetime of the `schedule_date` field (since the cron.trigger model has no other metadata than cronjob id and call_at).

This extra calculation step might not be necessary as ` action_put_in_queue` will generate a trigger for that datetime if `schedule_date` is set. So we can reasonably well expect that unless we have an uncaught error, the trigger is created.

File: addons/mass_mailing/models/mailing.py (hunk 16 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:288]: That would be the first case tested, with the frozen date of `datetime(2023, 2, 17, 9, 0)` we check for :
```
            self.assertEqual(mailing.next_departure, datetime(2023, 2, 17, 9, 0))
            self.assertIsNot(mailing.schedule_date, cron_job.nextcall)
            self.assertEqual(mailing.schedule_type, 'now')
            self.assertEqual(mailing.state, 'in_queue')
```
I.E. is the next_departure pointing to "now" which is also the trigger that should be created. But perhaps I am missing something? Perhaps a subtlety?

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py (hunk 14 above)
Comment (jorv-odoo) [reply addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:36]: Was not aware of this. Will push updated code once I have implemented any other code changes from the review comments!

File: addons/mass_mailing/models/mailing.py (hunk 15 above)
Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:280]: Then the follow-up question was, why do we assume it exists at the scheduled_date and not assume it exists when sending immediately?

Is that "extra minute" important to say it will be sent at 09:00 AM or at 09:01 AM?

I feel like if we assume the flow will create the triggers, then we should always assume it will.
And if we don't, then we check (all) created triggers.

Currently it's a bit of a mix of both.

Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:280]: You are right. The current version was the natural progression of the input from reth's comments.

But now that you point it out we could simplify the definition of `_compute_next_departure` and trust the flow, as any small time discrepancy would not have any real world impact.

What I can do is export the cron.trigger checks into the unit test, and simply assert if the relevant cron.trigger was indeed created. So if the cron.trigger flow ever changes in the future the tests should catch it.

Should I do that?

Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:280]: I think that's fair, but maybe keep your current version somewhere in case other reviewers don't agree :slightly_smiling_face:

File: addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py [hunk 17]
```diff
@@ -90,7 +84,72 @@ def test_mailing_next_departure(self):
 
             mailing.action_schedule()
 
+            cron_trigger = self.env['ir.cron.trigger'].sudo().search([
```
Comment (awa-odoo) [inline addons/mass_mailing/tests/test_mailing_mailing_schedule_date.py:87]: Please have a look at "capture_triggers" :slightly_smiling_face:

File: addons/mass_mailing/models/mailing.py (hunk 15 above)
Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:273]: @tde-banana-odoo Understandable!
So after awa's input lead me to an epiphany, I overhauled the whole thing. 
In summary:
- We can keep `_compute_next_departure` very simple, since the methods `action_launch` and
`action_schedule` should take care of the cron.trigger creation. Thus either we are gonna use the schedule_date or now() depending on if it is a scheduled mass mailing campaign or a "Send now" (i.e. no schedule_date set). Noo need to fetch the cron job id or cron time etc.
- I migrated all the check to see if the cron.triggers are created as expected into the unit tests. A long as those don't fail, there should be not reason to believe that our simplified version of `_compute_next_departure` should behave unexpectedly

Now if you don't like this approach I could revert the changes, but in my humble opinion this seems like a cleaner solution.

Comment (awa-odoo) [reply addons/mass_mailing/models/mailing.py:273]: In short:

When scheduling / posting, we currently create triggers (in action_put_in_queue).
So for the next departure we assume that triggers will be created and we display "now" if it's an immediate send and the schedule_date if it's scheduled.
And we add tests to ensure that triggers are correctly created.

(Correct me if I'm wrong here @jorv-odoo ).

If we don't want it to be that "simple", then we have to fetch all created triggers depending on shceduled_date and runtime of the cron, and it becomes much more complex.

Comment (tde-banana-odoo) [reply addons/mass_mailing/models/mailing.py:273]: Just to be sure: if triggers disappears, next sending possibility will anyway be the cron running itself, like once / day ?

Comment (jorv-odoo) [reply addons/mass_mailing/models/mailing.py:273]: The back-end side yes and per default that specific cron job runs once a day.
As long as the stats is in queue and schedule date < now, that mailing campaign should run
```
        mass_mailings = self.search([('state', 'in', ('in_queue', 'sending')), '|', ('schedule_date', '<', fields.Datetime.now()), ('schedule_date', '=', False)])
```

So if someone or something removes the trigger, the displayed next_departure would be slightly off time-wise depending on the cron job nextcall value.
